core/src/main/scala/org/ensime/core/Indexer.scala

```diff
@@ -73,16 +73,26 @@ class Indexer(
       import context.dispatcher
       val usages = index.findUsageLocations(fqn)
       val response = usages.map { usages =>
-        val results: List[LineSourcePosition] = usages.flatMap { u =>
+        val positions: List[LineSourcePosition] = usages.take(1000).flatMap { u =>
           val file = u.file
           file.map { f =>
             LineSourcePosition(
               EnsimeFile(Paths.get(new URI(f)).toString),
-              u.line.getOrElse(u.line.getOrElse(0))
+              u.line.getOrElse(0)
             )
           }
         }(collection.breakOut)
-        SourcePositions(results)
+        val files = positions.foldLeft[Set[EnsimeFile]](Set.empty) { (files, pos) =>
+          if (pos.line > 0) files + pos.file else files
+        }
+        val contents: Map[EnsimeFile, Array[String]] = files.map(f => f -> f.readAllLines.toArray)(collection.breakOut)
```

Comment:
Since a file may contain multiple references, looking up an index multiple times in an array would be faster compared to a list. (constant and linear). That's why I did that. Do you think we  would have better performance with keeping it as a list?
Though ofcourse a file may only have a single reference where converting to an array would have gone waste.